### PR TITLE
fix(autowatch): add guard for missing file error

### LINKED ISSUE #40: Error when adding new file with watcher
I have karma running with watcher enabled.

It throws an error when I create __new__ file (__Product-view.jsx__ is new file itself)
```shell
webpack: bundle is now INVALID.
ERROR [karma]: [Error: Path doesn't exist '/_karma_webpack_/specs/components/Product-view.jsx']
Error: Path doesn't exist '/_karma_webpack_/specs/components/Product-view.jsx'
    at MemoryFileSystem.readFileSync (/srv/node_modules/karma-webpack/node_modules/webpack-dev-middleware/node_modules/memory-fs/lib/MemoryFileSystem.js:82:10)
    at MemoryFileSystem.readFile (/srv/node_modules/karma-webpack/node_modules/webpack-dev-middleware/node_modules/memory-fs/lib/MemoryFileSystem.js:201:21)
    at doRead (/srv/node_modules/karma-webpack/index.js:157:26)
    at Plugin.readFile (/srv/node_modules/karma-webpack/index.js:161:3)
    at /srv/node_modules/karma-webpack/index.js:177:17
    at nextPreprocessor (/srv/node_modules/karma/lib/preprocessor.js:59:28)
    at /srv/node_modules/karma/lib/preprocessor.js:97:7
    at fs.js:336:14
    at /srv/node_modules/webpack/node_modules/watchpack/node_modules/chokidar/node_modules/readdirp/node_modules/graceful-fs/graceful-fs.js:104:5
    at /srv/node_modules/karma/node_modules/graceful-fs/graceful-fs.js:104:5
```
Attaching my karma.conf.js also:
```js
'use strict';

var webpack = require('webpack');
var path = require('path');
var ExtractTextPlugin = require("extract-text-webpack-plugin");

module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'sinon-chai'],

    // list of files / patterns to load in the browser
    files: [
      'specs/**/*.jsx'
    ],

    // list of files to exclude
    exclude: [
    ],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'specs/**/*.jsx': ["webpack"]
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    reporters: ['progress', 'coverage'],

    coverageReporter: {
        type : 'html',
        dir : 'coverage/'
    },

    // web server port
    port: 9876,

    // Timeout for capturing a browser (in ms).
    captureTimeout: 60 * 1e3,

    // to avoid DISCONNECTED messages
    browserDisconnectTimeout : 10000, // default 2000
    browserDisconnectTolerance : 1, // default 0
    browserNoActivityTimeout : 60 * 1e3, //default 10000

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_ERROR,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Include timeout because of rsync
    autoWatchBatchDelay: 2 * 1e3,

    // webpack options
    webpack: {
      cache: true,
      devtool: 'inline-source-map',
      watchDelay: 3 * 1e3,
      resolve: {
        extensions: ['', ".webpack.js", '.js', '.jsx'],
        modulesDirectories: ["node_modules", "bower_components"],
        alias: {
          'flux': path.join(__dirname, 'src/scripts/flux'),
          'src': path.join(__dirname, 'src/'),
          'routes': path.join(__dirname, 'src/scripts/routes'),
          'components': path.join(__dirname, 'src/scripts/components'),
          'scripts': path.join(__dirname, 'src/scripts'),
          'styles': path.join(__dirname, 'src/styles'),
          "kendo": "kendo-ui-webpack"
        }
      },
      module: {
        preLoaders: [
            {test: /(\.jsx)|(\.js)$/, exclude: /(node_modules|bower_components|specs)/, loader: 'isparta-instrumenter'},
        ],
        loaders: [
          {test: /(react)|(react\/addons)$/, loader: 'expose?React!imports?shim=es5-shim/es5-shim&sham=es5-shim/es5-sham'},
          {test: /\/jquery.js$/, loader: 'expose?jQuery'},
          {test: /react-tabs.*\.js$/, loader: '6to5-loader?experimental&runtime'},
          {test: /(\.jsx)|(\.js)$/, exclude: /node_modules/, loader: '6to5-loader?experimental&runtime'},
          {test: /\.css$/, loader:  'null'},
          {test: /\.(png|jpg|gif)$/, loader: 'null'},
          {test: /icon-.+\.(svg)$/, loader: 'null'},
          {test: /\.styl$/, loader: 'null'},
          {test: /\.woff.?$/, loader: "null" },
          {test: /fonts\/.*\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: "null"},
          {test: /\.json$/, loader: "json-loader"},
          {test: /bootstrap\/js\//, loader: 'imports?jQuery=jquery'},
          {test: /kendo\-ui\-core[\///].*\.js$/, loader: "imports?jQuery=jquery"},
        ]
      },
      plugins: [
        new webpack.PrefetchPlugin('react'),
        new webpack.PrefetchPlugin('jquery'),
        new webpack.PrefetchPlugin('lodash'),
        new webpack.PrefetchPlugin('fluxxor'),
        new webpack.PrefetchPlugin('react-router'),
        new webpack.ProvidePlugin({
          to5Runtime: "imports?global=>{}!exports?global.to5Runtime!6to5/runtime",
          React: 'react/addons',
          _: 'lodash',
          Router: 'react-router',
          Fluxxor: 'Fluxxor',
        }),
        new webpack.DefinePlugin({'global.BACKEND_HOST': '"http://localhost:8080/"'}),
        new webpack.DefinePlugin({'global.RESOURCE_PATH_PREFIX': '"http//localhost:8000"'}),
        new webpack.DefinePlugin({'process.env.NODE_ENV': '"development"', 'global.SIDE': '"frontend"'}),
        new webpack.ResolverPlugin(
          new webpack.ResolverPlugin.DirectoryDescriptionFilePlugin("bower.json", ["main"])
        )
      ]
    },

    webpackMiddleware: {
      stats: {
        // With console colors
        colors: true,
        // add the hash of the compilation
        hash: false,
        // add webpack version information
        version: false,
        // add timing information
        timings: true,
        // add assets information
        assets: false,
        // add chunk information
        chunks: false,
        // add built modules information to chunk information
        chunkModules: false,
        // add built modules information
        modules: false,
        // add also information about cached (not built) modules
        cached: false,
        // add information about the reasons why modules are included
        reasons: false,
        // add the source code of modules
        source: true,
        // add details to errors (like resolving log)
        errorDetails: true,
        // add the origins of chunks and chunk merging info
        chunkOrigins: true,
        // Add messages from child loaders
        children: false
      }
    },

  });
};
```

### PR DESCRIPTION
**Pull request by mewdriller**
Added a try/catch clause to guard against a "Path doesn't exist" error
coming from `MemoryFileSystem.prototype.readFileSync` when a new file is
added to the file-system that matches the files glob. Unfortunately, there
is not a custom type on the error, so we have to resort to string magic to
determine if that's the reason for the error.

Closes #40

### CODE REVIEW

**Comment by koggdal** [comment]
I can verify that this works for me at least. Thanks for fixing it! /ping @sokra

File: index.js [hunk 1]
```diff
@@ -156,8 +156,18 @@ Plugin.prototype.readFile = function(file, callback) {
 			middleware.fileSystem.readFile("/_karma_webpack_/" + file.replace(/\\/g, "/"), callback);
 		}
 	}
-	if(!this.waiting)
-		doRead();
+	if (!this.waiting) {
```
Comment (sokra) [inline index.js]: I wonder why this fix is needed. Shouldn't be `this.waiting` truthy until all files are available?

Comment (mewdriller) [inline index.js]: I would have assumed so, but from testing on my local machine (Windows 7 x64), `this.waiting` was evaluating to false whenever a new file was added that matched the glob during a watch.

I really don't understand the internals that are going on here. I simply used the logic that I would have expected to run (in the `this.waiting == true` branch) in the case that we observed the specific error that was getting thrown. It's entirely possible there's a simpler / more elegant way to resolve this issue. If you've got any suggestions or improvements let me know and I'd be happy to update things.

**Comment by Blystad** [comment]
+1 this fixed the problem for me.

**Comment by mewdriller** [comment]
@sokra: Are there any changes you'd like me to make to the PR or is it good as-is?

/cc @Dignifiedquire

**Comment by Dignifiedquire** [comment]
looks good to me :+1:

**Comment by cesarandreu** [comment]
+1 karma + webpack crash when I add a new file :(.

**Comment by Dignifiedquire** [comment]
ping @sokra

**Comment by mewdriller** [comment]
@sokra: Were you looking for this to be fixed in a different manner? If you've got another path you want me to pursue I can look into that.

**Comment by bassettsj** [comment]
:+1: Seems like a great fix for the time being. Should be merged in. :smile:

**Comment by le0pard** [comment]
+1

**Comment by SirCmpwn** [comment]
:+1:

**Comment by dignifiedquire** [comment]
ping @sokra

**Comment by MikaAK** [comment]
Is this still a thing I need to merge @dignifiedquire?

**Comment by dignifiedquire** [comment]
I don't think it's needed anymore. I haven't seen this error in a while anymore, so I would suggest closing it for now and if someone reproduces the issues with the latest versions looking into it again.

**Comment by Silviu-Marian** [comment]
This is still needed...

image

**Comment by d3viant0ne** [comment]
@mewdriller Evidently this is still an issue, can you please reopen a PR to the v1 branch so this issue can be properly addressed.

If I don't hear anything by tomorrow evening, I will address this issue myself.

**Comment by Silviu-Marian** [comment]
Let me know if you need a repository that reproduces the issue.

**Comment by d3viant0ne** [comment]
@Silviu-Marian - Given this was an issue under some contention in the past from what I can see above, that would be extremely helpful in getting this resolved quickly.

**Comment by Silviu-Marian** [comment]
https://github.com/Silviu-Marian/karma-webpack-pr69

**Pull request closed without merging**